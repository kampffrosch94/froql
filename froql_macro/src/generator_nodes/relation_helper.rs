use std::fmt::Write;

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct RelationHelperInfo {
    /// Index where the relationship cid is in the cid array for `old_var`
    pub cid_index: usize,
    pub column_index: usize,
    pub old_var: isize,
    pub new_var: isize,
    /// nr of this RelationHelper, used when generating variable name
    pub nr: usize,
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct UnrelationHelperInfo {
    /// Index where the relationship cid is in the cid array for `old_var`
    pub ty: String,
    pub flip_target: bool,
    pub old_var: isize,
    pub new_var: isize,
    /// nr of this RelationHelper, used when generating variable name
    pub nr: usize,
}

pub fn relation_helpers_init_and_set_col(
    prepend: &mut String,
    append: &mut String,
    helpers: &[RelationHelperInfo],
    unhelpers: &[UnrelationHelperInfo],
) {
    for helper in helpers {
        let old = helper.old_var;
        let nr = helper.nr;
        let column_index = helper.column_index;
        let cid_index = helper.cid_index;
        write!(
            prepend,
            "
let mut rel_helper_{nr} = ::froql::query_helper::RelationHelper::new
    (components_{old}[{cid_index}]);
"
        )
        .unwrap();
        write!(
            append,
            "
    rel_helper_{nr}.set_col(&a_ref.columns[col_indexes[{column_index}]]);
"
        )
        .unwrap();
    }

    for unhelper in unhelpers {
        let nr = unhelper.nr;
        let ty = &unhelper.ty;
        if unhelper.flip_target {
            write!(
                prepend,
                "
let mut unrel_helper_{nr} = ::froql::query_helper::UnrelationHelper::new
    (bk.get_component_id_unchecked(TypeId::of::<Relation<{ty}>>())
            .flip_target());
"
            )
            .unwrap();
        } else {
            write!(
                prepend,
                "
let mut unrel_helper_{nr} = ::froql::query_helper::UnrelationHelper::new
    (bk.get_component_id_unchecked(TypeId::of::<Relation<{ty}>>()));
"
            )
            .unwrap();
        }
        write!(
            append,
            "
    unrel_helper_{nr}.set_col(a_ref);
"
        )
        .unwrap();
    }
}

pub fn relation_helpers_set_rows(
    append: &mut String,
    helpers: &[RelationHelperInfo],
    unhelpers: &[UnrelationHelperInfo],
) {
    for helper in helpers {
        let nr = helper.nr;
        let var = helper.old_var;
        write!(
            append,
            "
        rel_helper_{nr}.set_row(bk, a_rows[{var}].0);
"
        )
        .unwrap();
    }
    for unhelper in unhelpers {
        let nr = unhelper.nr;
        let var = unhelper.old_var;
        write!(
            append,
            "
        unrel_helper_{nr}.set_row(bk, a_rows[{var}].0);
"
        )
        .unwrap();
    }
}
