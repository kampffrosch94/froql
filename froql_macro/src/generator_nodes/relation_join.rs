use super::relation_helper::{
    relation_helpers_init_and_set_col, relation_helpers_set_rows, RelationHelperInfo,
    UnrelationHelperInfo,
};
use super::types::{RelationConstraint, UnrelationConstraint};
use super::GeneratorNode;
use std::fmt::Write;
use std::ops::Range;

#[derive(Debug)]
pub struct RelationJoin {
    /// index of the component of `old` where the relation resides
    pub new: isize,
    pub new_components: Range<usize>,
    pub unequal_constraints: Vec<(isize, isize)>,
    /// RelationHelpers that constrain the new var
    pub rel_constraints: Vec<RelationConstraint>,
    /// UnrelationHelpers that constrain the new var
    pub unrel_constraints: Vec<UnrelationConstraint>,
    pub opt_components: Vec<(String, usize)>,
    /// RelationHelpers that depend on the new var
    pub new_relation_helpers: Vec<RelationHelperInfo>,
    /// UnrelationHelpers that depend on the new var
    pub new_unrelation_helpers: Vec<UnrelationHelperInfo>,
    /// the relationhelper that contains the Relation(old, new)
    pub new_helper_nr: usize,
}

impl GeneratorNode for RelationJoin {
    fn generate(&self, step: usize, prepend: &mut String, append: &mut String) -> usize {
        let new = self.new;
        let Range { start, end } = &self.new_components;
        let helper_nr = self.new_helper_nr;
        write!(
            append,
            "
// follow relation
{step} => {{
    const REL_VAR: usize = {new};
    const REL_VAR_COMPONENTS: ::std::ops::Range<usize> = {start}..{end};
    if let Some(id) = rel_helper_{helper_nr}.next_related() {{
        // get aid/row for entity in relation
        let (aid, arow) = bk.entities.get_archetype_unchecked(id);

        // if in target archetype set => go to next step
        if archetype_id_sets[REL_VAR].contains(&aid) {{
            let a_ref = &mut a_refs[REL_VAR];
            *a_ref = &bk.archetypes[aid.as_index()];
            a_ref.find_multiple_columns(
                &components_{new},
                &mut col_indexes[REL_VAR_COMPONENTS],
            );
            a_rows[REL_VAR] = arow;
"
        )
        .unwrap();

        // check constraints if there are any
        if self.unequal_constraints.is_empty()
            && self.rel_constraints.is_empty()
            && self.unrel_constraints.is_empty()
        {
            insert_optional_comps(prepend, append, &self.opt_components);
            relation_helpers_init_and_set_col(
                prepend,
                append,
                &self.new_relation_helpers,
                &self.new_unrelation_helpers,
            );
            relation_helpers_set_rows(
                append,
                &self.new_relation_helpers,
                &self.new_unrelation_helpers,
            );
            write!(
                append,
                "
            current_step += 1;"
            )
            .unwrap();
        } else {
            insert_checks(
                append,
                &self.unequal_constraints,
                &self.rel_constraints,
                &self.unrel_constraints,
            );
            append.push_str(
                "
            {
            } else {",
            );
            // handle optional components only in positive branch
            insert_optional_comps(prepend, append, &self.opt_components);
            relation_helpers_init_and_set_col(
                prepend,
                append,
                &self.new_relation_helpers,
                &self.new_unrelation_helpers,
            );
            relation_helpers_set_rows(
                append,
                &self.new_relation_helpers,
                &self.new_unrelation_helpers,
            );
            append.push_str(
                "
                current_step += 1;
            }",
            );
        }

        write!(
            append,
            "
        }}
    }} else {{
        current_step -= 1;
    }}
}}
"
        )
        .unwrap();
        return step + 1;
    }
}

pub fn insert_checks(
    append: &mut String,
    unequalities: &[(isize, isize)],
    rel_constraints: &[RelationConstraint],
    unrel_constraints: &[UnrelationConstraint],
) {
    append.push_str(
        r#"
            if"#,
    );

    let mut not_first = false;
    for (a, b) in unequalities {
        if not_first {
            write!(
                append,
                "
            ||"
            )
            .unwrap();
        }
        write!(
            append,
            "
                (::std::ptr::eq(a_refs[{a}], a_refs[{b}])
                 && a_rows[{a}] == a_rows[{b}])"
        )
        .unwrap();
        not_first = true;
    }
    for rc in rel_constraints {
        let helper_nr = &rc.helper_nr;
        let id = &rc
            .checked_invar
            .map(|it| format!("invar_{it}.id"))
            .unwrap_or_else(|| "id".to_string());
        if not_first {
            write!(
                append,
                "
            ||"
            )
            .unwrap();
        }
        write!(
            append,
            "
                !rel_helper_{helper_nr}.has_relation({id})
                "
        )
        .unwrap();
        not_first = true;
    }
    for rc in unrel_constraints {
        let helper_nr = &rc.helper_nr;
        let id = &rc
            .checked_invar
            .map(|it| format!("invar_{it}.id"))
            .unwrap_or_else(|| "id".to_string());
        if not_first {
            write!(
                append,
                "
            ||"
            )
            .unwrap();
        }
        write!(
            append,
            "
                !unrel_helper_{helper_nr}.satisfied({id})
                "
        )
        .unwrap();
        not_first = true;
    }
}

pub fn insert_optional_comps(
    prepend: &mut String,
    append: &mut String,
    opt_components: &[(String, usize)],
) {
    for (ty, id) in opt_components {
        write!(
            prepend,
            r#"
let opt_cid_{id} = world.get_component_id::<{ty}>();
let mut opt_col_{id} = None;"#
        )
        .unwrap();
        write!(
            append,
            r#"
            opt_col_{id} = a_ref.find_column_opt(opt_cid_{id});
"#
        )
        .unwrap();
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn relation_join_unequality() {
        let generator = RelationJoin {
            new: 2,
            new_components: 3..5,
            unequal_constraints: vec![(0, 2), (2, 1)],
            rel_constraints: vec![],
            unrel_constraints: vec![],
            opt_components: vec![],
            new_relation_helpers: vec![],
            new_unrelation_helpers: vec![],
            new_helper_nr: 0,
        };

        let mut prepend = String::new();
        let mut append = String::new();
        let r = generator.generate(3, &mut prepend, &mut append);
        assert_eq!(4, r);
        insta::assert_snapshot!(prepend, @"");
        insta::assert_snapshot!(append);
    }

    #[test]
    fn relation_join_constraint() {
        let generator = RelationJoin {
            new: 2,
            new_components: 3..5,
            unequal_constraints: vec![],
            rel_constraints: vec![RelationConstraint {
                helper_nr: 0,
                checked_invar: None,
            }],
            unrel_constraints: vec![],
            opt_components: vec![],
            new_relation_helpers: vec![],
            new_unrelation_helpers: vec![],
            new_helper_nr: 0,
        };

        let mut prepend = String::new();
        let mut append = String::new();
        let r = generator.generate(3, &mut prepend, &mut append);
        assert_eq!(4, r);
        insta::assert_snapshot!(prepend, @"");
        insta::assert_snapshot!(append);
    }

    #[test]
    fn relation_join_optional() {
        let generator = RelationJoin {
            new: 2,
            new_components: 3..5,
            unequal_constraints: vec![],
            rel_constraints: vec![],
            unrel_constraints: vec![],
            opt_components: vec![("OptA".into(), 0), ("OptB".into(), 1)],
            new_relation_helpers: vec![],
            new_unrelation_helpers: vec![],
            new_helper_nr: 0,
        };

        let mut prepend = String::new();
        let mut append = String::new();
        let r = generator.generate(3, &mut prepend, &mut append);
        assert_eq!(4, r);
        insta::assert_snapshot!(prepend, @r#"
        let opt_cid_0 = world.get_component_id::<OptA>();
        let mut opt_col_0 = None;
        let opt_cid_1 = world.get_component_id::<OptB>();
        let mut opt_col_1 = None;
        "#);
        insta::assert_snapshot!(append);
    }
}
