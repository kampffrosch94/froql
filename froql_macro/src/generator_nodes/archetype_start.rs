use crate::Checks;

use super::relation_helper::{
    relation_helpers_init_and_set_col, relation_helpers_set_rows, RelationHelperInfo,
    UnrelationHelperInfo,
};
use super::relation_join::{insert_checks, insert_optional_comps};
use super::GeneratorNode;
use std::fmt::Write;
use std::ops::Range;

#[derive(Debug)]
pub struct ArchetypeStart {
    pub var: isize,
    pub components: Range<usize>,
    pub opt_components: Vec<(String, usize)>,
    pub relation_helpers: Vec<RelationHelperInfo>,
    pub unrelation_helpers: Vec<UnrelationHelperInfo>,
    pub checks: Checks,
}

impl GeneratorNode for ArchetypeStart {
    fn generate(&self, step: usize, prepend: &mut String, append: &mut String) -> usize {
        let first = self.var;
        let Range { start, end } = &self.components;
        write!(
            append,
            "
{step} => {{
    const CURRENT_VAR: usize = {first};
    const CURRENT_VAR_COMPONENTS: ::std::ops::Range<usize> = {start}..{end};
    let next_index = &mut a_next_indexes[CURRENT_VAR];
    let archetype_ids = &archetype_id_sets[CURRENT_VAR];
    *next_index = next_index.wrapping_add(1);
    if *next_index >= archetype_ids.len() {{
        return None;
    }}
    let next_id = archetype_ids[*next_index];

    // gets rolled over to 0 by wrapping_add
    a_rows[CURRENT_VAR] = ArchetypeRow(u32::MAX);
    let a_ref = &mut a_refs[CURRENT_VAR];
    *a_ref = &bk.archetypes[next_id.as_index()];
    a_ref.find_multiple_columns(
        &components_{first},
        &mut col_indexes[CURRENT_VAR_COMPONENTS],
    );
    a_max_rows[CURRENT_VAR] = a_ref.entities.len() as u32;"
        )
        .unwrap();

        // handle optional components
        insert_optional_comps(prepend, append, &self.opt_components);
        relation_helpers_init_and_set_col(
            prepend,
            append,
            &self.relation_helpers,
            &self.unrelation_helpers,
        );

        write!(
            append,
            "
    current_step += 1;
}}
"
        )
        .unwrap();

        // get row from first archetype
        let next_step = step + 1;

        write!(
            append,
            "
// next row in archetype
{next_step} => {{
    const CURRENT_VAR: usize = {first};
    let row_counter = &mut a_rows[CURRENT_VAR].0;
    let max_row = a_max_rows[CURRENT_VAR];
    // rolls over to 0 for u32::MAX, which is our start value
    *row_counter = row_counter.wrapping_add(1);

    if *row_counter >= max_row {{
        current_step -= 1;
    }} else {{"
        )
        .unwrap();

        if self.checks.is_empty() {
            relation_helpers_set_rows(append, &self.relation_helpers, &self.unrelation_helpers);

            write!(
                append,
                "
        current_step += 1;"
            )
            .unwrap();
        } else {
            append.push_str(
                "
        let id = a_refs[CURRENT_VAR].entities[a_rows[CURRENT_VAR].as_index()];",
            );
            insert_checks(
                append,
                &self.checks.unequals,
                &self.checks.rel_constraints,
                &self.checks.unrel_constraints,
            );
            append.push_str(
                "
        {} else {",
            );
            relation_helpers_set_rows(append, &self.relation_helpers, &self.unrelation_helpers);

            append.push_str(
                "
            current_step += 1;
        }",
            );
        }

        write!(
            append,
            "
    }}
}}
"
        )
        .unwrap();

        return step + 2;
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn normal() {
        let gen = ArchetypeStart {
            var: 0,
            components: 0..2,
            opt_components: vec![],
            relation_helpers: vec![],
            unrelation_helpers: vec![],
            checks: Checks {
                unequals: vec![],
                rel_constraints: vec![],
                unrel_constraints: vec![],
            },
        };

        let mut prepend = String::new();
        let mut append = String::new();
        let r = gen.generate(0, &mut prepend, &mut append);
        assert_eq!(2, r);
        assert_eq!(prepend, "");
        insta::assert_snapshot!(append, @r#"
        0 => {
            const CURRENT_VAR: usize = 0;
            const CURRENT_VAR_COMPONENTS: ::std::ops::Range<usize> = 0..2;
            let next_index = &mut a_next_indexes[CURRENT_VAR];
            let archetype_ids = &archetype_id_sets[CURRENT_VAR];
            *next_index = next_index.wrapping_add(1);
            if *next_index >= archetype_ids.len() {
                return None;
            }
            let next_id = archetype_ids[*next_index];

            // gets rolled over to 0 by wrapping_add
            a_rows[CURRENT_VAR] = ArchetypeRow(u32::MAX);
            let a_ref = &mut a_refs[CURRENT_VAR];
            *a_ref = &bk.archetypes[next_id.as_index()];
            a_ref.find_multiple_columns(
                &components_0,
                &mut col_indexes[CURRENT_VAR_COMPONENTS],
            );
            a_max_rows[CURRENT_VAR] = a_ref.entities.len() as u32;
            current_step += 1;
        }

        // next row in archetype
        1 => {
            const CURRENT_VAR: usize = 0;
            let row_counter = &mut a_rows[CURRENT_VAR].0;
            let max_row = a_max_rows[CURRENT_VAR];
            // rolls over to 0 for u32::MAX, which is our start value
            *row_counter = row_counter.wrapping_add(1);

            if *row_counter >= max_row {
                current_step -= 1;
            } else {
                current_step += 1;
            }
        }
        "#);
    }

    #[test]
    fn check_unequal() {
        let gen = ArchetypeStart {
            var: 0,
            components: 0..2,
            opt_components: vec![],
            relation_helpers: vec![],
            unrelation_helpers: vec![],
            checks: Checks {
                unequals: vec![(0, 1)],
                rel_constraints: vec![],
                unrel_constraints: vec![],
            },
        };

        let mut prepend = String::new();
        let mut append = String::new();
        let r = gen.generate(0, &mut prepend, &mut append);
        assert_eq!(2, r);
        assert_eq!(prepend, "");
        insta::assert_snapshot!(append, @r#"
        0 => {
            const CURRENT_VAR: usize = 0;
            const CURRENT_VAR_COMPONENTS: ::std::ops::Range<usize> = 0..2;
            let next_index = &mut a_next_indexes[CURRENT_VAR];
            let archetype_ids = &archetype_id_sets[CURRENT_VAR];
            *next_index = next_index.wrapping_add(1);
            if *next_index >= archetype_ids.len() {
                return None;
            }
            let next_id = archetype_ids[*next_index];

            // gets rolled over to 0 by wrapping_add
            a_rows[CURRENT_VAR] = ArchetypeRow(u32::MAX);
            let a_ref = &mut a_refs[CURRENT_VAR];
            *a_ref = &bk.archetypes[next_id.as_index()];
            a_ref.find_multiple_columns(
                &components_0,
                &mut col_indexes[CURRENT_VAR_COMPONENTS],
            );
            a_max_rows[CURRENT_VAR] = a_ref.entities.len() as u32;
            current_step += 1;
        }

        // next row in archetype
        1 => {
            const CURRENT_VAR: usize = 0;
            let row_counter = &mut a_rows[CURRENT_VAR].0;
            let max_row = a_max_rows[CURRENT_VAR];
            // rolls over to 0 for u32::MAX, which is our start value
            *row_counter = row_counter.wrapping_add(1);

            if *row_counter >= max_row {
                current_step -= 1;
            } else {
                let id = a_refs[CURRENT_VAR].entities[a_rows[CURRENT_VAR].as_index()];
                    if
                        (::std::ptr::eq(a_refs[0], a_refs[1])
                         && a_rows[0] == a_rows[1])
                {} else {
                    current_step += 1;
                }
            }
        }
        "#);
    }
}
