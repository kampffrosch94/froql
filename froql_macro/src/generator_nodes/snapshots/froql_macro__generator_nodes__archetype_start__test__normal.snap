---
source: froql_macro/src/generator_nodes/archetype_start.rs
expression: append
---
0 => {
    const CURRENT_VAR: usize = 0;
    const CURRENT_VAR_COMPONENTS: ::std::ops::Range<usize> = 0..2;
    let next_index = &mut a_next_indexes[CURRENT_VAR];
    let archetype_ids = &archetype_id_sets[CURRENT_VAR];
    *next_index = next_index.wrapping_add(1);
    if *next_index >= archetype_ids.len() {
        return None;
    }
    let next_id = archetype_ids[*next_index];

    // gets rolled over to 0 by wrapping_add
    a_rows[CURRENT_VAR] = ArchetypeRow(u32::MAX);
    let a_ref = &mut a_refs[CURRENT_VAR];
    *a_ref = &bk.archetypes[next_id.as_index()];
    a_ref.find_multiple_columns(
        &components_0,
        &mut col_indexes[CURRENT_VAR_COMPONENTS],
    );
    a_max_rows[CURRENT_VAR] = a_ref.entities.len() as u32;
    current_step += 1;
}

// next row in archetype
1 => {
    const CURRENT_VAR: usize = 0;
    let row_counter = &mut a_rows[CURRENT_VAR].0;
    let max_row = a_max_rows[CURRENT_VAR];
    // rolls over to 0 for u32::MAX, which is our start value
    *row_counter = row_counter.wrapping_add(1);

    if *row_counter >= max_row {
        current_step -= 1;
    } else {
        current_step += 1;
    }
}
